dashboard/lib/expired_child_account_purger.rb

```diff
@@ -0,0 +1,212 @@
+require 'stringio'
+require 'cdo/aws/metrics'
+require 'cdo/aws/s3'
+require 'cdo/chat_client'
+require 'queries/child_account'
+
+# Scans for child accounts which should be hard-deleted because Code.org hasn't
+# received permission from their parents to create the account.
+#
+# Look at the `build_metrics` method to see which metrics are sent to
+# CloudWatch.
+#
+# Logs activity to Slack #cron-daily room.
+#
+# @see Technical Spec: Hard-deleting accounts
+# https://docs.google.com/document/d/15hkknuRlvGFbPuwlZssliMQTmykxM8_ajXB4yDOSPCA/edit
+# @see Account Purger Cloudwatch dashboard
+# TODO - Add link to the dashboard
+#
+class ExpiredChildAccountPurger
+  class SafetyConstraintViolation < RuntimeError; end
+
+  attr_reader :dry_run, :lock_out_date, :max_accounts_to_purge, :log
+  alias :dry_run? :dry_run
+
+  def initialize(options = {})
+    @dry_run = options[:dry_run].nil? ? false : options[:dry_run]
+    raise ArgumentError.new('dry_run must be boolean') unless [true, false].include? @dry_run
+
+    # The amount of time an account is allowed to be locked out before it is
+    # purged.
+    # The default is 7 days.
+    @lock_out_date = options[:lock_out_date] || 7.days.ago
+    raise ArgumentError.new('lock_out_date must be Time') unless @lock_out_date.is_a? Time
+
+    # Do nothing if more than this number of accounts would be purged in total.
+    # The motivation of this limit is to protect ourselves from bugs where we
+    # accidentally delete a large number of accounts we shouldn't be deleting.
+    # We may need to adjust this over time as activity increases on our site.
+    # Default to 1000
+    @max_accounts_to_purge = options[:max_accounts_to_purge] || 1000
```

Comment:
I wonder if we should lower the default on this. It seems like even as a sanity check 1000 seems high.

Reply:
I have lowered this to 200. Our current DeletedAccountPurger deletes about 100 accounts a day but usually has at least one 2-3k spike once a month. We can keep it at 200 to be safe and then adjust if we find we were being too conservative.